File: CS458_HW5/MISTA/CaseStudies/BlocksCaseStudy/CSharp/Blocks/Blocks/Block.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;

namespace Blocks
{
    class Block
    {
 /*       static void Main(string[] args)
        {
        Block block = new Block();
		block.Init();
		block.Unstack("1", "3");
		block.Putdown("1");
		block.Pickup("6");
		block.Stack("6", "1");
		Console.WriteLine("Final state: "+block.ToString());
        }
        */
        private List<string> ontables = new List<string>();
        private List<string> clears = new List<string>();
        private List<ON> ons = new List<ON>();
        private string holding = null;
        private bool handempty = true;

        public Block()
        {
        }

        public void Pickup(String a){
		if (ontables.Contains(a) && clears.Contains(a) && handempty){
			ontables.Remove(a);
			clears.Remove(a);
			handempty = false;
			holding = a;
		}
		else
            Console.WriteLine("Cannot pickup " + a);
	    }

        public void Putdown(String a){
		if (holding!=null){
			holding = null;
			ontables.Add(a);
			clears.Add(a);
			handempty = true;
		}
		else
            Console.WriteLine("Cannot put down " + a);
	    }

        public void Stack(String a, String b){
		if (a!=b && holding!=null && holding==a && clears.Contains(b)){	
			handempty=true;
			ons.Add(new ON(a,b));
			clears.Remove(b);
			clears.Add(a);
			holding = null;
		}
		else
            Console.WriteLine("Cannot stack " + a + " on " + b);
	    }

        public void Unstack(String a, String b){
		if (handempty && clears.Contains(a) && IsOn(a, b)){
			handempty = false;
			clears.Remove(a);
			RemoveOn(a, b);
			holding = a; 
			clears.Add(b);
		}
		else
            Console.WriteLine("Cannot unstack " + a + " from " + b);
	    }

        public List<string> GetOntables()
        {
            return ontables;
        }

        public bool IsOntable(string a)
        {
            return ontables.Contains(a);
        }

        public List<string> GetClears()
        {
            return clears;
        }

        public bool IsClear(String a)
        {
            return clears.Contains(a);
        }

        public List<ON> GetOns()
        {
            return ons;
        }

        public bool IsOn(string a, string b){
		ON on2 = new ON(a, b);
        for(int i=0; i < ons.Count; i++) 
			if (ons[i].Equals(on2))
				return true;
		return false;
	    }

        private void RemoveOn(string a, string b)
        {
            ON on2 = new ON(a, b);
            for (int i = ons.Count - 1; i >= 0; i--)
            {
                if (ons[i].Equals(on2))
                {
                    ons.RemoveAt(i);
                    return;
                }
            }
        }

        public string GetHolding()
        {
            return holding;
        }

        public bool IsHolding(string thing)
        {
            return holding!=null && holding.Equals(thing);
        }

        public bool IsHandempty()
        {
            return handempty;
        }

        public void SetHandempty(bool empty)
        {
            handempty = empty;
        }

        public override String ToString(){
		String str = "";
        for(int i=0; i < ontables.Count; i++) 
			str += "\n\tontable("+ontables[i]+")";
        for(int i=0; i < clears.Count; i++) 
			str += "\n\tclear("+clears[i]+")";
        for(int i=0; i < ons.Count; i++) 
			str += "\n\t"+ ons[i].ToString();
		if (holding!=null)
			str += "\n\tholding("+ holding+")";
		if (handempty)
			str += "\n\thandempty(\"r\")";
		return str;
	    }

        void Init()
        {
            ontables.Add("3");
            ontables.Add("6");
            clears.Add("1");
            clears.Add("6");
            ons.Add(new ON("1", "3"));
            handempty = true;
        }
    }

    public class ON
    {
        private string a, b;

        public ON(string a, string b)
        {
            this.a = a;
            this.b = b;
        }

        public bool Equals(ON on2)
        {
            return a.Equals(on2.a) && b.Equals(on2.b);
        }

        public override string ToString()
        {
            return "on(" + a + ", " + b + ")";
        }
    }
}

File: HW5/MISTA/examples/java/Blocks/blocks/Block.java
package blocks;

import java.util.ArrayList;

public class Block {

	private ArrayList<String> ontables = new ArrayList<String>();
	private ArrayList<String> clears = new ArrayList<String>();
	private ArrayList<ON> ons = new ArrayList<ON>();
	private String holding = null;
	private boolean handempty = true;
	
	public Block(){
	}
	
	public void pickup(String a){
		if (ontables.contains(a) && clears.contains(a) && handempty){
			ontables.remove(a);
			clears.remove(a);
			handempty = false;
			holding = a;
		}
		else 
			System.out.println("Cannot pickup "+a);
	}
	
	public void putdown(String a){
		if (holding!=null){
			holding = null;
			ontables.add(a);
			clears.add(a);
			handempty = true;
		}
		else 
			System.out.println("Cannot put down "+a);
	}
	
	public void stack(String a, String b){
		if (a!=b && holding!=null && holding==a && clears.contains(b)){	
			handempty=true;
			ons.add(new ON(a,b));
			clears.remove(b);
			clears.add(a);
			holding = null;
		}
		else 
			System.out.println("Cannot stack "+a +" on "+b);
	}

	public void unstack(String a, String b){
		if (handempty && clears.contains(a) && isOn(a, b)){
			handempty = false;
			clears.remove(a);
			removeOn(a, b);
			holding = a; 
			clears.add(b);
		}
		else
			System.out.println("Cannot unstack "+a+" from "+b);
	}

	public ArrayList<String> getOntables(){
		return ontables;
	}

	public boolean isOntable(String a){
		return ontables.contains(a);
	}
	
	public ArrayList<String> getClears(){
		return clears;
	}

	public boolean isClear(String a){
		return clears.contains(a);
	}

	public ArrayList<ON> getOns(){
		return ons;
	}
	public boolean isOn(String a, String b){
		ON on2 = new ON(a, b);
		for (ON on: ons)
			if (on.equals(on2))
				return true;
		return false;
	}

	private void removeOn(String a, String b){
		ON on2 = new ON(a, b);
		for (int i=ons.size()-1; i>=0; i--){
			if (ons.get(i).equals(on2)){
				ons.remove(i);
				return;
			}
		}
	}
	
	public String getHolding(){
		return holding;
	}

	public boolean isHolding(String thing){
		return holding!=null && holding.equalsIgnoreCase(thing);
	}
	
	public boolean isHandempty(){
		return handempty;
	}

	public void setHandempty(boolean empty){
		handempty = empty;
	}

	public String toString(){
		String str = "";
		for (String ontable: ontables)
			str += "\n\tontable("+ontable+")";
		for (String clear: clears)
			str += "\n\tclear("+clear+")";
		for (ON on: ons)
			str += "\n\t"+ on;
		if (holding!=null)
			str += "\n\tholding("+ holding+")";
		if (handempty)
			str += "\n\thandempty(\"r\")";
		return str;
	}
	
	void init(){
		ontables.add("3");
		ontables.add("6");
		clears.add("1");
		clears.add("6");
		ons.add(new ON("1","3"));
		handempty = true;
	}
	
	public static void main(String[] args){
		Block block = new Block();
		block.init();
		block.unstack("1", "3");
		block.putdown("1");
		block.pickup("6");
		block.stack("6", "1");
		System.out.println("Final state: "+block);
	}
}
